### ISSUE TITLE
Refactor kMipsCmpD part of AssembleArchBranch/Boolean?

### ISSUE THREAD

**Issue by paul99** [v8mips/v8mips#116]
With Benedikt's c24220c0 (https://codereview.chromium.org/850653002) and Balázs port of it (https://codereview.chromium.org/850733004) most of the Unordered comparisons have been removed.

They had been pushed into the arch ports back on October with this CL: [turbofan] Make VisitBranch and friends to backend specific. (https://codereview.chromium.org/651783003)

The portions under kMipsCmpD now look a bit different from the rest of the code, which gets Branch condition flags from FlagsConditionToConditionTst(), FlagsConditionToConditionOvf(), etc. We cannot do a simple function for CmpD, as it uses two values, for nan label ,and condition. 

I cannot quite get my head around this today, but we do need to review the importance of the unordered comparisons on the other arch's, and see if what we are doing can be simplified.

I suppose we should also review the whole branch mechanism looking for possible simplifications.

This should get done by either me or Dusan, but IMO now, the bugs take priority. I'll assign to myself for the moment, Dusan please feel free to grab this when time allows.

**Comment by paul99** [comment]
This was landed today as https://crrev.com/8f9a62d3120dbb346b421ee288458596cc82a9d9

Unfortunately, it had test failures on mjsunit/asm/embenchen/box2d, for instance here:
https://v8.mips.com/bbv8/builders/Sim%20-%20edge/builds/5394/steps/Check/logs/box2d
https://v8.mips.com/bbv8/builders/Sim%20-%20MIPS64%20-%20Debug%20-%20edge/builds/2329/steps/Check/logs/box2d

Here is the problem in gdb on mips64el:

```
# Fatal error in ../src/compiler/mips64/code-generator-mips64.cc, line 289
# unreachable code

Program received signal SIGABRT, Aborted.
0x00007fff90605286 in __pthread_kill ()
(gdb) bt
#0  0x00007fff90605286 in __pthread_kill ()
#1  0x00007fff89c4142f in pthread_kill ()
#2  0x00007fff921d8b53 in abort ()
#3  0x00000001009e8018 in v8::base::OS::Abort () at ../src/base/platform/platform-posix.cc:233
#4  0x00000001009e1d39 in V8_Fatal (file=0x100a59c00 "../src/compiler/mips64/code-generator-mips64.cc", line=289, format=0x1009ee360 "unreachable code") at ../src/base/logging.cc:119
#5  0x00000001009c77e6 in v8::internal::compiler::(anonymous namespace)::FlagsConditionToConditionCmpD (predicate=@0x7fff5fbfbeb7, condition=v8::internal::compiler::kUnsignedGreaterThan) at ../src/compiler/mips64/code-generator-mips64.cc:289
#6  0x00000001009c74a8 in v8::internal::compiler::CodeGenerator::AssembleArchBoolean (this=0x7fff5fbfc4e0, instr=0x1028bfc18, condition=v8::internal::compiler::kUnsignedGreaterThan) at ../src/compiler/mips64/code-generator-mips64.cc:1001
#7  0x00000001002db750 in v8::internal::compiler::CodeGenerator::AssembleInstruction (this=0x7fff5fbfc4e0, instr=0x1028bfc18) at ../src/compiler/code-generator.cc:192
#8  0x00000001002db143 in v8::internal::compiler::CodeGenerator::GenerateCode (this=0x7fff5fbfc4e0) at ../src/compiler/code-generator.cc:73
#9  0x000000010036bd7a in v8::internal::compiler::GenerateCodePhase::Run (this=0x7fff5fbfc760, data=0x7fff5fbfd400, temp_zone=0x1265006d0, linkage=0x7fff5fbfd278) at ../src/compiler/pipeline.cc:692
#10 0x000000010036a99e in v8::internal::compiler::Pipeline::Run<v8::internal::compiler::GenerateCodePhase, v8::internal::compiler::Linkage*> (this=0x7fff5fbfdd18, arg_0=0x7fff5fbfd278) at ../src/compiler/pipeline.cc:345
#11 0x0000000100368a6d in v8::internal::compiler::Pipeline::GenerateCode (this=0x7fff5fbfdd18, linkage=0x7fff5fbfd278) at ../src/compiler/pipeline.cc:1060
#12 0x00000001003681da in v8::internal::compiler::Pipeline::GenerateCode (this=0x7fff5fbfdd18) at ../src/compiler/pipeline.cc:915
#13 0x00000001003e2243 in v8::internal::OptimizedCompileJob::CreateGraph (this=0x7fff5fbfdf58) at ../src/compiler.cc:429
```

FlagsConditionToConditionCmpD() was passed a condition `kUnsignedGreaterThan`, which is not in this new function, but was in the old AssembleArchBoolean(). Also missing is kUnsignedGreaterThanOrEqual. True for mips32 as well.

Clearly, the comparison could be inverted to use the conditionals that are provided, but I don't know how to tell the higher layer of the compiler to do that.

@dusmil, can you please take a look, and fix and resubmit? (You likely already know, but I saw Jakob or someone suggest that when re-submitting a reverted change in a new CL, upload the original (reverted) code as patchset 1, so that the fix is clear.)

**Comment by dusmil** [comment]
@paul99:  Patch resubmitted. Higher layers already do not use GT, GE conditions, but in our other optimization for comparison with zero the condition can be negated.

https://codereview.chromium.org/905493002/

**Comment by paul99** [comment]
Landed as c03a5ca1 MIPS:[turbofan] Reland "Improve unordered comparisons for boolean materialization."

Thanks. Closing.
